### ISSUE TITLE
Ambiguity about read-only CSRs and illegal instruction exceptions

### ISSUE THREAD

**Issue by benjaminselfridge** [riscv/riscv-isa-manual#221]
I sent the following email to the formal spec group, regarding what I believe is an ambiguity in the spec regarding writing to read-only CSRs:

At the bottom of page 5 of the privilege spec, it says "Attempts to
access a CSR without appropriate privilege level or to write a
read-only register also raise illegal instruction exceptions."

My question is -- what precisely constitutes a CSR read/write access
vs. a read-only access? Obviously expansions of the pseudoinstruction
CSRR should be considered read-only, which at the very least means
that CSRRS is read-only whenever the source register is x0. Is that
the only example of a read-only CSR access? Other potential examples
of read-only behavior are:

- CSRRS rd, csr where x[rd] == 0 and rd is any register id
- CSRRS rd, csr where CSRs[csr] | x[rs1] == CSRs[csr] (i.e. the value
that would be written is the same as what's already there
- *ANY* CSR instruction where whatever value would be written (if any)
to the CSR happens to evaluate to what is already in there

An example:

Suppose that x[1] = 0x00000000 and x[2] = 0x0000FFFF (on an RV32
system). Suppose also that marchid (a read-only register) is equal to
0xFFFFFFFF.

My understanding is that the following instruction:

CSRRS x3, marchid, x0

would *not* raise an illegal instruction exception, because x0 is
hardwired to zero. In this instance, x0 gets ORed with the original
value of marchid, which produces that same value back, and under this
particular circumstance we consider this to *not* be a write (under
the terminology of the passage I quoted from page 5). Otherwise, there
would be no way to read this register at all!

Now, my question is, would either of the following raise illegal
instruction exceptions:

CSRRS x3, marchid, x1
CSRRS x3, marchid, x2

In other words: since marchid | x[rs1] == marchid and marchid | x[rs2]
== marchid, both of the above instructions could be conceived of as
legal, since they are not attempting to write *new* values back into
marchid. So, are they illegal, or not? Is CSRRS illegal on a read-only
CSR whenever the source operand is anything other than x0, or simply
when the result that *WOULD* be written into the CSR ends up being the
same as what is already in there?

**Comment by benjaminselfridge** [comment]
I've just realized that this ambiguity has been clarified since I last built the user-level manual. This passage from 8.1 clears it up: "Note that if rs1 specifies a register holding a zero value other than x0, the instruction will still attempt to write the unmodified value back to the CSR and will cause any attendant side effects."
